# Assignment #2: Data Exploration

In this assignment, you will demonstrate your knowledge of the Python and Pandas skills we've learned so far. These include:

- Getting an overview of your data
- Extracting columns
- Removing duplicates from your data
- Creating a subset of your data by matching a string
- Getting a random sample of your data
- Sorting your data 
- Getting fundamental statistical information of a series, such as the mean and median

You will not need to formulate or answer a research question for this assignment. You simply need to demonstrate your ability to perform specific operations on the data set you have selected. In the next assignment, your midterm, you will create a narrative exploration of the data using many of the same methods, but in such a way as to use narrative along with data analysis to answer a research question.

You may use the data set you imported in assignment #1, or you may choose another data set. That means that you can reuse code you wrote in Assignment #1 to import the data.

In the below notebook, I will ask for you to perform a task. Each task will be followed by one or more blank code cells. Use as many cells as you need to complete the task. If you know some other way to complete the task with Python and/or Pandas that is not the way we learned in class, that is also acceptable. Feel free to use Google searches to refresh your memory as to how to perform each task. You will also likely wish to consult the class notebooks, which can be found in [our class repository on GitHub](https://github.com/sha256rma/foundations-of-data-science).

If your data set does not have a needed form of data for a task, write code that would perform the task if a column with that data existed, and provide an explanation as a comment or markdown cell.

You may receive partial credit for some incomplete or incorrect answers. Please feel free to add comments about your thought process, which will make it more likely that you will receive partial credit.

The last cell is a bonus. You will not be penalized for not attempting or completing it.

-----

Import Pandas and read your data in as a data frame. Assign the data frame to a variable, such as `df`.

In [7]:
import pandas

In [9]:
df = pandas.read_csv('mxmh_survey_results.csv')

Make the data frame visible by placing the data frame variable in a cell by itself.

In [11]:
df

Timestamp   Age Primary streaming service  Hours per day  \
0     8/27/2022 19:29:02  18.0                   Spotify            3.0   
1     8/27/2022 19:57:31  63.0                   Pandora            1.5   
2     8/27/2022 21:28:18  18.0                   Spotify            4.0   
3     8/27/2022 21:40:40  61.0             YouTube Music            2.5   
4     8/27/2022 21:54:47  18.0                   Spotify            4.0   
..                   ...   ...                       ...            ...   
731  10/30/2022 14:37:28  17.0                   Spotify            2.0   
732   11/1/2022 22:26:42  18.0                   Spotify            1.0   
733   11/3/2022 23:24:38  19.0   Other streaming service            6.0   
734   11/4/2022 17:31:47  19.0                   Spotify            5.0   
735    11/9/2022 1:55:20  29.0             YouTube Music            2.0   

    While working Instrumentalist Composer         Fav genre Exploratory  \
0             Yes             Yes      Yes             Latin         Yes   
1             Yes              No       No              Rock         Yes   
2              No              No       No  Video game music          No   
3             Yes              No      Yes              Jazz         Yes   
4             Yes              No       No               R&B         Yes   
..            ...             ...      ...               ...         ...   
731           Yes             Yes       No              Rock         Yes   
732           Yes             Yes       No               Pop         Yes   
733           Yes              No      Yes               Rap         Yes   
734           Yes             Yes       No         Classical          No   
735           Yes              No       No           Hip hop         Yes   

    Foreign languages  ...  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0                 Yes  ...        Sometimes  Very frequently            Never   
1                  No  ...        Sometimes           Rarely  Very frequently   
2                 Yes  ...            Never           Rarely           Rarely   
3                 Yes  ...        Sometimes            Never            Never   
4                  No  ...  Very frequently  Very frequently            Never   
..                ...  ...              ...              ...              ...   
731               Yes  ...            Never           Rarely  Very frequently   
732               Yes  ...            Never            Never        Sometimes   
733                No  ...        Sometimes        Sometimes           Rarely   
734                No  ...            Never            Never            Never   
735               Yes  ...  Very frequently  Very frequently  Very frequently   

    Frequency [Video game music] Anxiety Depression Insomnia  OCD  \
0                      Sometimes     3.0        0.0      1.0  0.0   
1                         Rarely     7.0        2.0      2.0  1.0   
2                Very frequently     7.0        7.0     10.0  2.0   
3                          Never     9.0        7.0      3.0  3.0   
4                         Rarely     7.0        2.0      5.0  9.0   
..                           ...     ...        ...      ...  ...   
731                        Never     7.0        6.0      0.0  9.0   
732                    Sometimes     3.0        2.0      2.0  5.0   
733                       Rarely     2.0        2.0      2.0  2.0   
734                    Sometimes     2.0        3.0      2.0  1.0   
735                       Rarely     2.0        2.0      2.0  5.0   

    Music effects    Permissions  
0             NaN  I understand.  
1             NaN  I understand.  
2       No effect  I understand.  
3         Improve  I understand.  
4         Improve  I understand.  
..            ...            ...  
731       Improve  I understand.  
732       Improve  I understand.  
733       Improve  I understand.  
734       Improve  I understand.  
735       Impr

What columns are in your data frame? Use a function or method that shows all the columns.

In [20]:
df.columns

Index(['Timestamp', 'Age', 'Primary streaming service', 'Hours per day',
       'While working', 'Instrumentalist', 'Composer', 'Fav genre',
       'Exploratory', 'Foreign languages', 'BPM', 'Frequency [Classical]',
       'Frequency [Country]', 'Frequency [EDM]', 'Frequency [Folk]',
       'Frequency [Gospel]', 'Frequency [Hip hop]', 'Frequency [Jazz]',
       'Frequency [K pop]', 'Frequency [Latin]', 'Frequency [Lofi]',
       'Frequency [Metal]', 'Frequency [Pop]', 'Frequency [R&B]',
       'Frequency [Rap]', 'Frequency [Rock]', 'Frequency [Video game music]',
       'Anxiety', 'Depression', 'Insomnia', 'OCD', 'Music effects',
       'Permissions'],
      dtype='object')

Write Python code to output the number of rows in your data frame.

In [30]:
rows = len(df)
rows

736

Write code to output a column from your data frame as a series. (Extract a column.)

In [33]:
df['Age']

0      18.0
1      63.0
2      18.0
3      61.0
4      18.0
       ... 
731    17.0
732    18.0
733    19.0
734    19.0
735    29.0
Name: Age, Length: 736, dtype: float64

### Read the example below to **count** duplicates based on specific columns (e.g., 'Name' and 'Age')

Check duplicates based on 'Name' and 'Age' columns (modify the subset for your own columns):

`duplicates_subset = df.duplicated(subset=['Name', 'Age'])`

Count the number of duplicates in the subset:

`num_duplicates_subset = duplicates_subset.sum()`

Display no. of duplicate rows based on your column(s):

`print(f"\nNumber of duplicate rows based on 'Name' and 'Age': {num_duplicates_subset}")`


After reading the above, modify the code snippets to check for duplicate columns in your own data frame

In [38]:
duplicates_subset = df.duplicated(subset=['Fav genre', 'Age'])

In [41]:
num_duplicates_subset = duplicates_subset.sum()

In [43]:
print(f"\nNumber of duplicate rows based on 'Fav genre' and 'Age': {num_duplicates_subset}")


Number of duplicate rows based on 'Fav genre' and 'Age': 456


If any duplicates exist within a column, remove them with the `.drop_duplicates()` function

In [50]:
df.drop_duplicates(subset=['Fav genre', 'Age'], inplace=True)

Write Python code to output the number of rows in your data frame.

In [57]:
df

Timestamp   Age Primary streaming service  Hours per day  \
0    8/27/2022 19:29:02  18.0                   Spotify            3.0   
1    8/27/2022 19:57:31  63.0                   Pandora            1.5   
2    8/27/2022 21:28:18  18.0                   Spotify            4.0   
3    8/27/2022 21:40:40  61.0             YouTube Music            2.5   
4    8/27/2022 21:54:47  18.0                   Spotify            4.0   
..                  ...   ...                       ...            ...   
702  10/3/2022 10:53:28  37.0                   Spotify            2.0   
707  10/3/2022 16:43:32  26.0                   Spotify            1.0   
709   10/4/2022 0:48:34  35.0                   Spotify            4.0   
714  10/4/2022 19:33:42  29.0               Apple Music            1.0   
728  10/30/2022 7:24:08  14.0   Other streaming service            7.0   

    While working Instrumentalist Composer         Fav genre Exploratory  \
0             Yes             Yes      Yes             Latin         Yes   
1             Yes              No       No              Rock         Yes   
2              No              No       No  Video game music          No   
3             Yes              No      Yes              Jazz         Yes   
4             Yes              No       No               R&B         Yes   
..            ...             ...      ...               ...         ...   
702           Yes              No       No              Folk         Yes   
707            No              No       No           Country         Yes   
709           Yes             Yes      Yes           Hip hop         Yes   
714            No              No       No           Country         Yes   
728           Yes             Yes       No           Country         Yes   

    Foreign languages  ...  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0                 Yes  ...        Sometimes  Very frequently            Never   
1                  No  ...        Sometimes           Rarely  Very frequently   
2                 Yes  ...            Never           Rarely           Rarely   
3                 Yes  ...        Sometimes            Never            Never   
4                  No  ...  Very frequently  Very frequently            Never   
..                ...  ...              ...              ...              ...   
702                No  ...           Rarely           Rarely        Sometimes   
707                No  ...            Never            Never        Sometimes   
709                No  ...  Very frequently  Very frequently        Sometimes   
714                No  ...           Rarely           Rarely  Very frequently   
728                No  ...        Sometimes        Sometimes  Very frequently   

    Frequency [Video game music] Anxiety Depression Insomnia  OCD  \
0                      Sometimes     3.0        0.0      1.0  0.0   
1                         Rarely     7.0        2.0      2.0  1.0   
2                Very frequently     7.0        7.0     10.0  2.0   
3                          Never     9.0        7.0      3.0  3.0   
4                         Rarely     7.0        2.0      5.0  9.0   
..                           ...     ...        ...      ...  ...   
702                    Sometimes     6.0        4.0      5.0  0.0   
707                       Rarely     4.0        1.0      1.0  6.0   
709              Very frequently     7.0        8.0      6.0  5.0   
714                        Never     9.0        3.0      2.0  0.0   
728                       Rarely     7.0        3.0      1.0  2.0   

    Music effects    Permissions  
0             NaN  I understand.  
1             NaN  I understand.  
2       No effect  I understand.  
3         Improve  I understand.  
4         Improve  I understand.  
..            ...            ...  
702       Improve  I understand.  
707       Improve  I understand.  
709       Improve  I understand.  
714     No effect  I understand.  
728       Improve  I unde

Compare the new value with the first time you outputted the number of rows above. Is it more, less, or the same?

There were 456 less rows 

Create a subset of the data that matches a specific string in a column. That is, extract all rows of the original data frame that contain a specific string in one of the columns. Save the resulting data frame to a variable. (Add more cells if needed)

In [67]:
df['Fav genre'] == 'Pop'

0      False
1      False
2      False
3      False
4      False
       ...  
702    False
707    False
709    False
714    False
728    False
Name: Fav genre, Length: 280, dtype: bool

In [72]:
df.loc[df['Fav genre'] == 'Pop', ['Age', 'Fav genre']]

Age Fav genre
15   17.0       Pop
23   18.0       Pop
36   21.0       Pop
49   19.0       Pop
53   23.0       Pop
106  41.0       Pop
107  20.0       Pop
111  16.0       Pop
132  32.0       Pop
139  25.0       Pop
143  53.0       Pop
181  35.0       Pop
195  26.0       Pop
201  22.0       Pop
224  31.0       Pop
245  38.0       Pop
285  27.0       Pop
305  56.0       Pop
324  59.0       Pop
367  43.0       Pop
369  10.0       Pop
379  15.0       Pop
387  14.0       Pop
456  49.0       Pop
494  74.0       Pop
573  29.0       Pop
668  24.0       Pop

Output the length of the resulting data frame (the subset of your data for which the conditional was true).

In [85]:
rows = len(df.loc[df['Fav genre'] == 'Pop', ['Age', 'Fav genre']])
rows

27

Output the first ten rows of your data set, the last ten rows of your data set, and a random ten rows of your data set (a sample of your data).

In [90]:
df.head(10)

Timestamp   Age Primary streaming service  Hours per day  \
0   8/27/2022 19:29:02  18.0                   Spotify            3.0   
1   8/27/2022 19:57:31  63.0                   Pandora            1.5   
2   8/27/2022 21:28:18  18.0                   Spotify            4.0   
3   8/27/2022 21:40:40  61.0             YouTube Music            2.5   
4   8/27/2022 21:54:47  18.0                   Spotify            4.0   
5   8/27/2022 21:56:50  18.0                   Spotify            5.0   
7   8/27/2022 22:18:59  21.0                   Spotify            1.0   
8   8/27/2022 22:33:05  19.0                   Spotify            6.0   
10  8/27/2022 22:51:15  18.0                   Spotify            3.0   
11  8/27/2022 23:00:32  19.0             YouTube Music            8.0   

   While working Instrumentalist Composer         Fav genre Exploratory  \
0            Yes             Yes      Yes             Latin         Yes   
1            Yes              No       No              Rock         Yes   
2             No              No       No  Video game music          No   
3            Yes              No      Yes              Jazz         Yes   
4            Yes              No       No               R&B         Yes   
5            Yes             Yes      Yes              Jazz         Yes   
7            Yes              No       No             K pop         Yes   
8            Yes              No       No              Rock          No   
10           Yes             Yes       No           Country         Yes   
11           Yes              No       No               EDM         Yes   

   Foreign languages  ...  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0                Yes  ...        Sometimes  Very frequently            Never   
1                 No  ...        Sometimes           Rarely  Very frequently   
2                Yes  ...            Never           Rarely           Rarely   
3                Yes  ...        Sometimes            Never            Never   
4                 No  ...  Very frequently  Very frequently            Never   
5                Yes  ...  Very frequently  Very frequently  Very frequently   
7                Yes  ...        Sometimes           Rarely            Never   
8                 No  ...            Never            Never  Very frequently   
10                No  ...           Rarely            Never           Rarely   
11                No  ...           Rarely        Sometimes           Rarely   

   Frequency [Video game music] Anxiety Depression Insomnia  OCD  \
0                     Sometimes     3.0        0.0      1.0  0.0   
1                        Rarely     7.0        2.0      2.0  1.0   
2               Very frequently     7.0        7.0     10.0  2.0   
3                         Never     9.0        7.0      3.0  3.0   
4                        Rarely     7.0        2.0      5.0  9.0   
5                         Never     8.0        8.0      7.0  7.0   
7                        Rarely     5.0        3.0      5.0  3.0   
8                         Never     2.0        0.0      0.0  0.0   
10                        Never     7.0        7.0      4.0  7.0   
11                       Rarely     1.0        0.0      0.0  1.0   

   Music effects    Permissions  
0            NaN  I understand.  
1            NaN  I understand.  
2      No effect  I understand.  
3        Improve  I understand.  
4        Improve  I understand.  
5        Improve  I understand.  
7        Improve  I understand.  
8        Improve  I understand.  
10     No effect  I understand.  
11       Improve  I understand.  

[10 rows x 33 columns]

In [92]:
df.tail(10)

Timestamp   Age          Primary streaming service  \
669  9/15/2022 15:30:41  20.0                            Spotify   
673  9/18/2022 16:23:29  27.0                            Spotify   
679  9/19/2022 23:53:37  36.0                      YouTube Music   
686  9/23/2022 23:06:29  18.0  I do not use a streaming service.   
695  9/28/2022 17:25:48  89.0                            Spotify   
702  10/3/2022 10:53:28  37.0                            Spotify   
707  10/3/2022 16:43:32  26.0                            Spotify   
709   10/4/2022 0:48:34  35.0                            Spotify   
714  10/4/2022 19:33:42  29.0                        Apple Music   
728  10/30/2022 7:24:08  14.0            Other streaming service   

     Hours per day While working Instrumentalist Composer  Fav genre  \
669            1.0           Yes              No       No      K pop   
673           10.0           Yes             Yes      Yes       Jazz   
679            3.0           Yes              No       No    Country   
686            1.0           Yes              No       No  Classical   
695           24.0           Yes             Yes      Yes        Rap   
702            2.0           Yes              No       No       Folk   
707            1.0            No              No       No    Country   
709            4.0           Yes             Yes      Yes    Hip hop   
714            1.0            No              No       No    Country   
728            7.0           Yes             Yes       No    Country   

    Exploratory Foreign languages  ...  Frequency [R&B]  Frequency [Rap]  \
669         Yes               Yes  ...        Sometimes            Never   
673         Yes               Yes  ...        Sometimes           Rarely   
679          No               Yes  ...        Sometimes            Never   
686          No                No  ...           Rarely           Rarely   
695          No                No  ...        Sometimes  Very frequently   
702         Yes                No  ...           Rarely           Rarely   
707         Yes                No  ...            Never            Never   
709         Yes                No  ...  Very frequently  Very frequently   
714         Yes                No  ...           Rarely           Rarely   
728         Yes                No  ...        Sometimes        Sometimes   

    Frequency [Rock] Frequency [Video game music] Anxiety Depression Insomnia  \
669           Rarely                        Never     8.0        7.0      1.0   
673        Sometimes                    Sometimes     6.0       10.0      8.0   
679            Never                        Never     1.0        0.0      1.0   
686           Rarely                       Rarely     2.0        0.0      0.0   
695           Rarely                        Never     0.0        0.0      0.0   
702        Sometimes                    Sometimes     6.0        4.0      5.0   
707        Sometimes                       Rarely     4.0        1.0      1.0   
709        Sometimes              Very frequently     7.0        8.0      6.0   
714  Very frequently                        Never     9.0        3.0      2.0   
728  Very frequently                       Rarely     7.0        3.0      1.0   

     OCD Music effects    Permissions  
669  1.0     No effect  I understand.  
673  1.0       Improve  I understand.  
679  0.0       Improve  I understand.  
686  5.0     No effect  I understand.  
695  0.0     No effect  I understand.  
702  0.0       Improve  I understand.  
707  6.0       Improve  I understand.  
709  5.0       Improve  I understand.  
714  0.0     No effect  I understand.  
728  2.0       Improve  I understand.  

[10 rows x 33 columns]

In [95]:
df.sample(10)

Timestamp   Age          Primary streaming service  \
18   8/27/2022 23:40:55  17.0                            Spotify   
324   8/29/2022 5:41:25  59.0                      YouTube Music   
490   9/1/2022 19:44:33  71.0  I do not use a streaming service.   
52   8/28/2022 12:52:19  23.0                        Apple Music   
412   8/30/2022 5:16:13  26.0                            Spotify   
516    9/2/2022 6:20:43  24.0                      YouTube Music   
436  8/31/2022 12:53:28  14.0                            Spotify   
619  9/12/2022 13:33:58  26.0                            Spotify   
128  8/28/2022 17:27:12  24.0                            Spotify   
46   8/28/2022 12:15:02  36.0                            Spotify   

     Hours per day While working Instrumentalist Composer  Fav genre  \
18           24.00           Yes              No       No        Rap   
324           2.00           Yes              No       No        Pop   
490           0.25           Yes              No       No     Gospel   
52            1.00            No              No       No        Rap   
412           6.00           Yes              No       No        EDM   
516           1.00           Yes              No       No       Folk   
436           9.00           Yes              No       No        R&B   
619           3.00           Yes             Yes      Yes       Jazz   
128           1.00            No              No       No      Metal   
46            1.00            No             Yes      Yes  Classical   

    Exploratory Foreign languages  ...  Frequency [R&B]  Frequency [Rap]  \
18          Yes                No  ...        Sometimes  Very frequently   
324         Yes                No  ...  Very frequently           Rarely   
490          No               Yes  ...           Rarely            Never   
52           No               Yes  ...           Rarely  Very frequently   
412          No               Yes  ...           Rarely           Rarely   
516          No               Yes  ...           Rarely        Sometimes   
436         Yes               Yes  ...  Very frequently  Very frequently   
619         Yes               Yes  ...        Sometimes           Rarely   
128         Yes                No  ...        Sometimes        Sometimes   
46           No                No  ...            Never            Never   

    Frequency [Rock] Frequency [Video game music] Anxiety Depression Insomnia  \
18   Very frequently                        Never     7.0        5.0      0.0   
324        Sometimes                        Never     1.0        1.0      3.0   
490           Rarely                        Never     1.0        0.0      0.0   
52   Very frequently                       Rarely    10.0        6.0      3.0   
412            Never                       Rarely    10.0       10.0      8.0   
516  Very frequently              Very frequently    10.0       10.0      7.0   
436           Rarely                       Rarely    10.0        8.0      9.0   
619  Very frequently                       Rarely     6.0        8.0      0.0   
128  Very frequently                       Rarely     6.0        4.0      5.0   
46         Sometimes                        Never     8.0        9.0      3.0   

      OCD Music effects    Permissions  
18    3.0       Improve  I understand.  
324   0.0       Improve  I understand.  
490   0.0       Improve  I understand.  
52   10.0     No effect  I understand.  
412   2.0     No effect  I understand.  
516   8.0       Improve  I understand.  
436   5.0       Improve  I understand.  
619   0.0       Improve  I understand.  
128   0.0       Improve  I understand.  
46    0.0     No effect  I understand.  

[10 rows x 33 columns]

## Bonus

You may want to refer to the [Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).

Sort your data frame by a particular column and output the result.

In [111]:
df_sorted = df.sort_values('Fav genre')

In [113]:
print(df_sorted)

              Timestamp   Age          Primary streaming service  \
392  8/29/2022 19:07:50  14.0                            Spotify   
46   8/28/2022 12:15:02  36.0                            Spotify   
384  8/29/2022 17:58:47  12.0                      YouTube Music   
390  8/29/2022 18:56:27  54.0  I do not use a streaming service.   
92   8/28/2022 14:49:18  24.0                            Spotify   
..                  ...   ...                                ...   
174  8/28/2022 19:29:08  16.0  I do not use a streaming service.   
438  8/31/2022 14:28:33  38.0                        Apple Music   
461   9/1/2022 15:24:02  34.0  I do not use a streaming service.   
42   8/28/2022 11:55:54  37.0                      YouTube Music   
192  8/28/2022 20:20:52  22.0                            Spotify   

     Hours per day While working Instrumentalist Composer         Fav genre  \
392           2.00           Yes             Yes       No         Classical   
46            1.00       

Pick a column in your data set with numeric data (integers or floats). Output the mean (average) and median of that column.

In [117]:
mean_value = df['Anxiety'].mean()
median_value = df['Anxiety'].median()

In [119]:
print("Mean:", mean_value)
print("Median:", median_value)

Mean: 5.582142857142857
Median: 6.0
